stackit/internal/resources/argus/instance/actions.go

```diff
@@ -429,60 +343,17 @@ func (r Resource) Delete(ctx context.Context, req resource.DeleteRequest, resp *
 		resp.Diagnostics.AddError("can't perform deletion", "argus instance id is unknown or null")
 	}
 
-	deleted := types.Bool{Value: false}
-	if err := helper.RetryContext(ctx, common.DURATION_30M, r.delete(ctx, &deleted, &state)); err != nil {
+	c := r.Provider.Client()
+	_, process, err := c.Argus.Instances.Delete(ctx, state.ProjectID.Value, state.ID.Value)
+	if err != nil {
 		resp.Diagnostics.AddError("failed to delete instance", err.Error())
-		return
 	}
 
-	resp.State.RemoveResource(ctx)
-}
-
-func (r *Resource) delete(ctx context.Context, deleted *types.Bool, state *Instance) func() *helper.RetryError {
-	c := r.Provider.Client()
-
-	return func() *helper.RetryError {
-		if !deleted.Value {
-			_, err := c.Argus.Instances.Delete(ctx, state.ProjectID.Value, state.ID.Value)
-			if err != nil {
-				if strings.Contains(err.Error(), "instance is not in the right state") ||
-					strings.Contains(err.Error(), "Could not delete instance") {
-					return helper.RetryableError(err)
-				}
-				if common.IsNonRetryable(err) {
-					return helper.NonRetryableError(err)
-				}
-				return helper.RetryableError(err)
-			}
-			deleted.Value = true
-		}
-
-		// give API a couple of seconds to process
-		time.Sleep(2 * time.Second)
-
-		// verify deletion
-		res, err := c.Argus.Instances.List(ctx, state.ProjectID.Value)
-		if err != nil {
-			if strings.Contains(err.Error(), http.StatusText(http.StatusNotFound)) {
-				return nil
-			}
-			return helper.RetryableError(err)
-		}
-		for _, i := range res.Instances {
-			if i.Instance != state.ID.Value {
-				continue
-			}
-			if i.Status == consts.ARGUS_INSTANCE_STATUS_DELETE_SUCCEEDED {
-				return nil
-			}
-			if i.Status == consts.ARGUS_INSTANCE_STATUS_CREATE_SUCCEEDED {
-				deleted.Value = false
-			}
-			return helper.RetryableError(fmt.Errorf("deletion incomplete. instance %s has status %s", i.Name, i.Status))
-
-		}
-		return nil
+	if _, err := process.Wait(); err != nil {
+		resp.Diagnostics.AddError("failed verify instance deletion", err.Error())
```

Comment:
Shouldn't there be a `return` here? Or maybe it's worth to add a comment why there's no `return` or the `RemoveResource` call can be put with a `defer` at the beginning to ensure that it gets called even in case of error?

Reply:
fixed